components/feature/search/src/main/java/mozilla/components/feature/search/middleware/SearchMiddleware.kt

```diff
@@ -0,0 +1,168 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.search.middleware
+
+import android.content.Context
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.launch
+import mozilla.components.browser.state.action.BrowserAction
+import mozilla.components.browser.state.action.InitAction
+import mozilla.components.browser.state.action.SearchAction
+import mozilla.components.browser.state.search.RegionState
+import mozilla.components.browser.state.search.SearchEngine
+import mozilla.components.browser.state.state.BrowserState
+import mozilla.components.feature.search.storage.BundledSearchEnginesStorage
+import mozilla.components.feature.search.storage.CustomSearchEngineStorage
+import mozilla.components.feature.search.storage.SearchMetadataStorage
+import mozilla.components.lib.state.Middleware
+import mozilla.components.lib.state.MiddlewareContext
+import mozilla.components.lib.state.Store
+import java.util.Locale
+
+/**
+ * [Middleware] implementation for loading and saving [SearchEngine]s whenever the state changes.
+ */
+class SearchMiddleware(
+    context: Context,
+    private val customStorage: CustomStorage = CustomSearchEngineStorage(context),
+    private val bundleStorage: BundleStorage = BundledSearchEnginesStorage(context),
+    private val metadataStorage: MetadataStorage = SearchMetadataStorage(context),
+    private val ioDispatcher: CoroutineDispatcher = Dispatchers.IO
+) : Middleware<BrowserState, BrowserAction> {
+    override fun invoke(
+        context: MiddlewareContext<BrowserState, BrowserAction>,
+        next: (BrowserAction) -> Unit,
+        action: BrowserAction
+    ) {
+        when (action) {
+            InitAction -> init(context.store)
+            is SearchAction.SetRegionAction -> loadRegionSearchEngines(context.store, action.regionState)
+            is SearchAction.UpdateCustomSearchEngineAction -> saveCustomSearchEngine(action)
+            is SearchAction.RemoveCustomSearchEngineAction -> removeCustomSearchEngine(action)
+            is SearchAction.SetDefaultSearchEngineAction -> updateDefaultSearchEngine(action)
+        }
+
+        next(action)
+    }
+
+    private fun init(
+        store: Store<BrowserState, BrowserAction>
+    ) {
+        loadMetadata(store)
+        loadCustomSearchEngines(store)
+    }
+
+    private fun loadMetadata(
+        store: Store<BrowserState, BrowserAction>
+    ) = GlobalScope.launch(ioDispatcher) {
+        val id = metadataStorage.getDefaultSearchEngineId()
+        if (id != null) {
+            store.dispatch(SearchAction.SetDefaultSearchEngineAction(id))
```

Comment:
I have question here to make sure I understood the flow. When `SearchMiddleware` receives the `InitAction` it will read the default search engine ID here and dispatch it to the store via `SetDefaultSearchEngineAction`.

When `RegionMiddleware` receives the `InitAction` it starts determining the region and dispatches to the store via `SetRegionAction`, which is "intercepted" above here in `SearchMiddleware` to trigger loading search engines from the bundle storage.

Is that correct? It might be worth documenting this init flow in the KDocs of this Middleware.

Reply:
Yeah, to summarize:

**InitAction**
- _RegionMiddleware_ - Determines region of user and dispatches `SetRegionAction`
- _SearchMiddleware_ - Loads custom search engines and dispatches `SetCustomSearchEngines`. And loads default search engine ID and dispatches `SetDefaultSearchEngineAction`. (Both only happens if the user has picked a default, or has added custom search engines).

**SetRegionAction**
- _SearchMiddleware_ - Loads the bundled search engines for that region and dispatches `SetRegionSearchEngines`.


We could have done all search engine loading when observing `SetRegionAction` in `SearchMiddleware`, but it felt unnecessary to wait since the other values are independent from the region. Something to consider though. It may make it easier to, for example, force a load of the default search engine of the user if it isn't found in either search engine list (e.g. the user may have a new _home_ region where this search engine is not in the configuration).

Reply:
👍